### PR TITLE
[SPARK-23914][SQL][follow-up] refactor ArrayUnion

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR refactors `ArrayUnion` based on [this suggestion](https://github.com/apache/spark/pull/21103#discussion_r205668821).
1. Generate optimized code for all of the primitive types except `boolean`
1. Generate code using `ArrayBuilder` or `ArrayBuffer`
1. Leave only a generic path in the interpreted path

## How was this patch tested?

Existing tests

### CODE REVIEW

**Comment by holdensmagicalunicorn** [comment]
@kiszk, thanks! I am a bot who has found some folks who might be able to help with the review:@ueshin, @rxin and @colorant

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -3698,230 +3767,162 @@ object ArraySetLike {
...
+        }
+
         s"""
-           |$openHashSet $hs = new $openHashSet$postFix($classTag);
-           |boolean $foundNullElement = false;
+           |$openHashSet $hashSet = new $openHashSet$hsPostFix($classTag);
+           |$declareNullTrackVariables
+           |int $size = 0;
+           |$arrayBuilderClass $builder =
+           |  ($arrayBuilderClass)$arrayBuilder.make($arrayBuilderClassTag);
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: nit: new `$arrayBuilderClass()` should work?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -3698,230 +3767,162 @@ object ArraySetLike {
...
-        val arrays = ctx.freshName("arrays")
+        val nullElementIndex = ctx.freshName("nullElementIndex")
+        val builder = ctx.freshName("builder")
         val array = ctx.freshName("array")
+        val arrays = ctx.freshName("arrays")
         val arrayDataIdx = ctx.freshName("arrayDataIdx")
+        val openHashSet = classOf[OpenHashSet[_]].getName
+        val classTag = s"scala.reflect.ClassTag$$.MODULE$$.$hsTypeName()"
+        val hashSet = ctx.freshName("hashSet")
+        val arrayBuilder = "scala.collection.mutable.ArrayBuilder"
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: nit: `classOf[mutable.ArrayBuilder[_]].getName`?

**Comment by ueshin** [comment]
I'll revisit after the conflict is fixed.

**Comment by kiszk** [comment]
I see. Now, I am rebasing and resolving conflicts.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 3]
```diff
@@ -3767,230 +3767,160 @@ object ArraySetLike {
...
+        val nullElementIndex = ctx.freshName("nullElementIndex")
+        val builder = ctx.freshName("builder")
         val array = ctx.freshName("array")
+        val arrays = ctx.freshName("arrays")
         val arrayDataIdx = ctx.freshName("arrayDataIdx")
+        val openHashSet = classOf[OpenHashSet[_]].getName
+        val classTag = s"scala.reflect.ClassTag$$.MODULE$$.$hsTypeName()"
+        val hashSet = ctx.freshName("hashSet")
+        val arrayBuilder = classOf[mutable.ArrayBuilder[_]].getName
+        val arrayBuilderClass = s"$arrayBuilder$$of$ptName"
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: This is not needed any more?

Comment (kiszk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: I think that we still need this to create an intermediate result array. The array is allocated at L3907 as other functions do.

Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Oh, sorry, I misread.

**Comment by ueshin** [comment]
LGTM.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**